cmd/migrateblobs/main.go

```diff
@@ -0,0 +1,151 @@
+// discarded.  The first time this command is executed, all the entities are
+// updated.  Subsequent runs migrates missing destination blobs.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+
+	"github.com/juju/loggo"
+	"github.com/juju/utils/parallel"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/goose.v2/client"
+	"gopkg.in/goose.v2/errors"
+	"gopkg.in/goose.v2/identity"
+	"gopkg.in/goose.v2/swift"
+	"gopkg.in/mgo.v2"
+
+	"gopkg.in/juju/charmstore.v5-unstable/config"
+)
+
+var (
+	logger        = loggo.GetLogger("migrateblobs")
+	loggingConfig = flag.String("logging-config", "INFO", "specify log levels for modules e.g. <root>=TRACE")
+)
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "usage: %s [options] <source config path> <destination config path>\n", filepath.Base(os.Args[0]))
+		flag.PrintDefaults()
+		fmt.Fprintf(os.Stderr, "Source config path is used to it read charmstore data. Destination config path is used only for writing new blobs.\n")
+		os.Exit(2)
+	}
+	flag.Parse()
+	if flag.NArg() != 2 {
+		flag.Usage()
+	}
+	if *loggingConfig != "" {
+		if err := loggo.ConfigureLoggers(*loggingConfig); err != nil {
+			fmt.Fprintf(os.Stderr, "cannot configure loggers: %v", err)
+			os.Exit(1)
+		}
+	}
+	if err := run(flag.Arg(0), flag.Arg(1)); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func run(srcConfPath, dstConfPath string) error {
+	logger.Infof("reading configuration")
+	srcConf, err := config.Read(srcConfPath)
+	if err != nil {
+		return errgo.Notef(err, "cannot read config file %q", srcConfPath)
+	}
+	dstConf, err := config.Read(dstConfPath)
+	if err != nil {
+		return errgo.Notef(err, "cannot read config file %q", dstConfPath)
+	}
+
+	logger.Infof("connecting to mongo")
+	session, err := mgo.Dial(srcConf.MongoURL)
+	if err != nil {
+		return errgo.Notef(err, "cannot dial mongo at %q", srcConf.MongoURL)
+	}
+	defer session.Close()
+	db := session.DB("juju")
+
+	cred := &identity.Credentials{
+		URL:        dstConf.SwiftAuthURL,
+		User:       dstConf.SwiftUsername,
+		Secrets:    dstConf.SwiftSecret,
+		Region:     dstConf.SwiftRegion,
+		TenantName: dstConf.SwiftTenant,
+	}
+	client := client.NewClient(cred, dstConf.SwiftAuthMode.Mode, nil)
+	client.SetRequiredServiceTypes([]string{"object-store"})
+	dst := swift.New(client)
+
+	logger.Infof("migrating entity blobs")
+	if err := migrate(db.GridFS("entitystore"), dst, dstConf.SwiftBucket); err != nil {
+		return errgo.Notef(err, "cannot migrate entity blobs")
+	}
+
+	logger.Infof("done")
+	return nil
+}
+
+func migrate(gridfs *mgo.GridFS, sc *swift.Client, dstContainerName string) error {
+	iter := gridfs.Find(nil).Sort("-uploadDate").Iter()
+	defer iter.Close()
+	run := parallel.NewRun(10)
+	var file *mgo.GridFile
+	var counter int32
+	var alreadyExistsCounter int32
+	for gridfs.OpenNext(iter, &file) {
+		// OpenNext close the file before going to the next one
+		file1 := file
+		file = nil
```

Comment:
// Zero the file pointer so that OpenNext doesn't close it on the next iteration.

?